go/exorelay/index.go

```diff
@@ -0,0 +1,143 @@
+package exorelay
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Originate/exocom/go/exosocket"
+	"github.com/Originate/exocom/go/structs"
+)
+
+// ExoRelay is the mid level API to Exocom
+type ExoRelay struct {
+	closed               bool
+	exoSocket            *exosocket.ExoSocket
+	messageHandlers      MessageHandlerMapping
+	messageHandlersMutex sync.RWMutex
+	replyChannels        ReplyChannelMapping
+	replyChannelsMutex   sync.RWMutex
+}
+
+// NewExoRelay creates a new ExoRelay instance with the given config
+func NewExoRelay(config exosocket.Config) *ExoRelay {
+	return &ExoRelay{
+		closed:          false,
+		exoSocket:       exosocket.NewExoSocket(config),
+		replyChannels:   ReplyChannelMapping{},
+		messageHandlers: MessageHandlerMapping{},
+	}
+}
+
+// Connect connects to ExoCom and starts listening for messages
+func (e *ExoRelay) Connect() error {
+	err := e.exoSocket.Connect()
+	if err != nil {
+		return err
+	}
+	go e.listenForMessages()
+	return nil
+}
+
+// Close disconnects from ExoCom
+func (e *ExoRelay) Close() error {
+	if e.closed {
+		return nil
+	}
+	err := e.exoSocket.Close()
+	e.closed = true
+	return err
+}
+
+// RegisterHandler adds a handler for the given message name
+func (e *ExoRelay) RegisterHandler(messageName string, handler MessageHandler) {
+	e.messageHandlersMutex.Lock()
+	e.messageHandlers[messageName] = handler
+	e.messageHandlersMutex.Unlock()
+}
+
+// Send sends a message
+func (e *ExoRelay) Send(messageOptions exosocket.MessageOptions) (*structs.Message, error) {
+	return e.exoSocket.Send(messageOptions)
```

Comment:
Is this method necessary? Go provides infrastructure for composition already.

Reply:
We don't want to expose all of exosocket's methods. Only a couple

Reply:
Interesting. Maybe explain in a comment why we deviate from the standard Go way here, so that the next person doesn't refactor this to using composition?